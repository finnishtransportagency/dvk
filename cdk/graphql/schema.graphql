schema {
  query: Query
  mutation: Mutation
}

type Query {
  fairwayCards(status: [Status!]): [FairwayCard!]!
  fairwayCard(id: String!, version: String!): FairwayCard
  fairwayCardVersion(id: String!, version: String!): FairwayCard
  fairwayCardLatest(id: String!): FairwayCard
  fairwayCardPublic(id: String!): FairwayCard
  fairwayCardPreview(id: String!, version: String!): FairwayCard
  safetyEquipmentFaults: [SafetyEquipmentFault!]!
  marineWarnings: [MarineWarning!]!
  currentUser: CurrentUser
  pilotPlaces: [PilotPlace!]!
  harbors: [Harbor!]!
  harbor(id: String!, version: String!): Harbor
  harborVersion(id: String!, version: String!): Harbor
  harborLatest(id: String!): Harbor
  harborPublic(id: String!): Harbor
  harborPreview(id: String!, version: String!): Harbor
  fairways: [Fairway!]!
  fairwayCardsAndHarbors: [FairwayCardOrHarbor!]!
  mareographs: [Mareograph!]!
}

type Mutation {
  saveFairwayCard(card: FairwayCardInput!): FairwayCard
  saveHarbor(harbor: HarborInput!): Harbor
  uploadPicture(picture: PictureUploadInput!): Boolean
}

input PictureUploadInput {
  cardId: String!
  id: String!
  base64Data: String!
  contentType: String!
}

enum Operation {
  CREATE
  UPDATE
}

enum OperationError {
  CARD_ID_MISSING
  CARD_ALREADY_EXIST
  CARD_NOT_EXIST
  HARBOR_ID_MISSING
  HARBOR_ALREADY_EXIST
  HARBOR_NOT_EXIST
  NO_PERMISSIONS
  INVALID_INPUT
}

input FairwayCardInput {
  id: String!
  version: String!
  operation: Operation!
  name: TextInput!
  n2000HeightSystem: Boolean!
  group: String!
  status: Status!
  fairwayIds: [Float!]!
  primaryFairwayId: [SelectedFairwayInput!]
  secondaryFairwayId: [SelectedFairwayInput!]
  additionalInfo: TextInput
  lineText: TextInput
  designSpeed: TextInput
  generalInfo: TextInput
  anchorage: TextInput
  navigationCondition: TextInput
  iceCondition: TextInput
  attention: TextInput
  speedLimit: TextInput
  visibility: TextInput
  vesselRecommendation: TextInput
  mareographs: [Int!]
  windRecommendation: TextInput
  trafficService: TrafficServiceInput
  harbors: [String!]
  pictures: [PictureInput!]
  pilotRoutes: [Int!]
  temporaryNotifications: [TemporaryNotificationInput!]
}

input PictureInput {
  id: String!
  sequenceNumber: Int
  orientation: Orientation!
  rotation: Float
  scaleWidth: String
  scaleLabel: String
  text: String
  lang: String
  groupId: Float
  modificationTimestamp: Float
  legendPosition: String
}

enum Orientation {
  PORTRAIT
  LANDSCAPE
}

input TrafficServiceInput {
  pilot: PilotInput
  vts: [VTSInput]
  tugs: [TugInput]
}

input TugInput {
  name: TextInput!
  phoneNumber: [String!]
  fax: String
  email: String
}

input VHFInput {
  name: TextInput
  channel: String!
}

input VTSInput {
  name: TextInput!
  phoneNumber: String
  email: [String]
  vhf: [VHFInput]
}

input PilotInput {
  email: String
  phoneNumber: String
  fax: String
  internet: String
  extraInfo: TextInput
  places: [PilotPlaceInput!]
}

input PilotPlaceInput {
  id: Float!
  pilotJourney: String
}

input HarborInput {
  id: String!
  version: String!
  n2000HeightSystem: Boolean!
  operation: Operation!
  status: Status!
  phoneNumber: [String]
  fax: String
  email: String
  internet: String
  name: TextInput!
  company: TextInput
  extraInfo: TextInput
  harborBasin: TextInput
  cargo: TextInput
  quays: [QuayInput]
  geometry: GeometryInput!
}

enum Status {
  DRAFT
  PUBLIC
  REMOVED
}

input GeometryInput {
  lat: String!
  lon: String!
}

input TextInput {
  fi: String!
  sv: String!
  en: String!
}

input QuayInput {
  name: TextInput
  length: String
  sections: [SectionInput]
  extraInfo: TextInput
  geometry: GeometryInput
}

input SectionInput {
  name: String
  depth: String
  geometry: GeometryInput
}

input TemporaryNotificationInput {
  content: TextInput!
  startDate: String!
  endDate: String
}

input SelectedFairwayInput {
  id: Int!
  sequenceNumber: Int!
}

type FairwayCardOrHarbor {
  id: String!
  version: String!
  type: ContentType!
  name: Text!
  group: String
  fairwayIds: [Int!]
  n2000HeightSystem: Boolean!
  status: Status!
  creator: String
  modifier: String
  modificationTimestamp: Float
  temporaryNotifications: [TemporaryNotification]
}

enum ContentType {
  CARD
  HARBOR
}

type CurrentUser {
  uid: String!
  name: String!
  roles: [String!]!
}

type FairwayCard {
  id: String!
  version: String!
  name: Text!
  n2000HeightSystem: Boolean!
  group: String
  status: Status
  creator: String
  creationTimestamp: Float
  modifier: String
  modificationTimestamp: Float
  fairwayIds: String
  fairways: [Fairway!]!
  additionalInfo: Text
  lineText: Text
  designSpeed: Text
  generalInfo: Text
  anchorage: Text
  navigationCondition: Text
  iceCondition: Text
  attention: Text
  speedLimit: Text
  visibility: Text
  vesselRecommendation: Text
  mareographs: [Mareograph!]
  windRecommendation: Text
  trafficService: TrafficService
  harbors: [Harbor!]
  pictures: [Picture!]
  pilotRoutes: [PilotRoute!]
  temporaryNotifications: [TemporaryNotification!]
}

type Picture {
  id: String!
  sequenceNumber: Int
  orientation: Orientation!
  rotation: Float
  text: String
  lang: String
  scaleLabel: String
  scaleWidth: String
  groupId: Float
  modificationTimestamp: Float
  legendPosition: String
}

type TrafficService {
  pilot: Pilot
  vts: [VTS]
  tugs: [Tug]
}

type Tug {
  name: Text
  phoneNumber: [String]
  fax: String
  email: String
}

type VHF {
  name: Text
  channel: Int
}

type VTS {
  name: Text
  phoneNumber: String
  email: [String]
  vhf: [VHF]
}

type Pilot {
  email: String
  phoneNumber: String
  fax: String
  internet: String
  extraInfo: Text
  places: [PilotPlace!]
}

type PilotPlace {
  id: Float!
  name: Text!
  pilotJourney: Float
  geometry: GeometryPoint!
}

type Harbor {
  id: String!
  version: String
  status: Status
  n2000HeightSystem: Boolean
  creator: String
  creationTimestamp: Float
  modifier: String
  modificationTimestamp: Float
  quays: [Quay]
  name: Text
  company: Text
  phoneNumber: [String]
  fax: String
  email: String
  internet: String
  extraInfo: Text
  harborBasin: Text
  cargo: Text
  geometry: GeometryPoint
}

type Quay {
  name: Text
  length: Float
  sections: [Section]
  extraInfo: Text
  geometry: GeometryPoint
}

type Section {
  name: String
  depth: Float
  geometry: GeometryPoint
}

type Fairway {
  id: Int!
  name: Text
  primary: Boolean
  primarySequenceNumber: Int
  secondary: Boolean
  secondarySequenceNumber: Int
  startText: String
  endText: String
  lightingCode: String
  lighting: Text
  typeCode: String
  type: Text
  owner: String
  area: Text
  sizing: Sizing
  classifications: [Classification!]
  sizingVessels: [SizingVessel!]
  navigationLines: [NavigationLine!]
  areas: [Area!]
  restrictionAreas: [RestrictionArea!]
  prohibitionAreas: [ProhibitionArea!]
  boardLines: [Boardline!]
  turningCircles: [TurningCircle!]
}

type Classification {
  type: String
  fairwayClassCode: String
  fairwayClass: Text
}

type AreaFairway {
  fairwayId: Int!
  status: Int
  line: Int
  sizingSpeed: Float
  sizingSpeed2: Float
  sequenceNumber: Int
}

type Area {
  id: Int!
  name: String
  depth: Float
  draft: Float
  referenceLevel: String
  n2000depth: Float
  n2000draft: Float
  n2000ReferenceLevel: String
  direction: Float
  journalNumber: String
  verificationDate: String
  owner: String
  additionalInformation: String
  typeCode: Int
  type: String
  notationCode: Float
  notation: String
  operationStatusCode: Float
  operationStatus: String
  operationTypeCode: Float
  operationType: String
  operationDirectionCode: Float
  operationDirection: String
  fairways: [AreaFairway!]
  geometry: GeometryPolygon
}

type RestrictionArea {
  id: Int!
  types: [RestrictionType!]
  value: Float
  presenter: String
  journalNumber: String
  verificationDate: String
  modificationDate: String
  startDate: String
  endDate: String
  status: String
  source: String
  location: String
  municipality: String
  exception: String
  geometry: GeometryPolygon
  fairways: [RestrictionFairway!]
}

type RestrictionType {
  code: String
  text: String
}

type RestrictionFairway {
  fairwayId: Int!
}

type ProhibitionArea {
  id: Int!
  extraInfo: Text
  fairway: ProhibitionFairway!
}

type ProhibitionFairway {
  fairwayId: Int!
  name: Text
}

type BoardlineFairway {
  fairwayId: Int!
}

type Boardline {
  id: Int!
  direction: Float
  geometry: GeometryLine
  fairways: [BoardlineFairway!]
}

type TurningCircleFairway {
  fairwayId: Int!
}

type TurningCircle {
  id: Int!
  geometry: GeometryLine
  diameter: Int
  fairways: [TurningCircleFairway!]
}

type SizingVessel {
  typeCode: String
  type: String
  length: Float
  width: Float
  draft: Float
  size: Float
  bodyFactor: Float
}

type Sizing {
  normalTurningCircle: Float
  minimumTurningCircle: Float
  normalWidth: Float
  minimumWidth: Float
  additionalInformation: String
  mareograph: String
  reserveWater: String
}

type NavigationLine {
  id: Int!
  depth: Float
  draft: Float
  referenceLevel: String
  n2000depth: Float
  n2000draft: Float
  n2000ReferenceLevel: String
  direction: Float
  length: Float
  owner: String
  verificationDate: String
  journalNumber: String
  additionalInformation: String
  type: String
  typeCode: String
  geometry: GeometryLine
  fairways: [NavigationLineFairway!]
}

type NavigationLineFairway {
  fairwayId: Int!
  status: Int
  line: Int
}

type SafetyEquipmentFault {
  id: Int!
  name: Text
  equipmentId: Int!
  typeCode: Int!
  type: Text
  recordTime: Float!
  geometry: GeometryPoint!
}

type GeometryPolygon {
  type: String
  coordinates: [[[Float]]]
}

type GeometryLine {
  type: String
  coordinates: [[Float]]
}

type GeometryPoint {
  type: String
  coordinates: [Float]
}

type Text {
  fi: String
  sv: String
  en: String
}

type MarineWarning {
  id: Int!
  number: Int!
  area: Text!
  type: Text!
  location: Text!
  description: Text!
  startDateTime: Float
  endDateTime: Float
  dateTime: Float!
  notifier: String!
  equipmentId: Float
  lineId: Float
  areaId: Float
}

type PilotRoute {
  id: Int!
  name: String
}

type TemporaryNotification {
  content: Text
  startDate: String
  endDate: String
}

type Mareograph {
  id: Int!
  name: String
  dateTime: Float
  waterLevel: Float
  n2000WaterLevel: Float
  calculated: Boolean
}
