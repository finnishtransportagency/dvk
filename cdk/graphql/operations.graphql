query findFairways {
  fairways {
    ...fragment FairwayFragment on Fairway {
      id
      nameFI
      nameSV
    }
  }
}

query findFairway($id: Int) {
  fairway(id: $id) {
    ...Fairway
  }
}
