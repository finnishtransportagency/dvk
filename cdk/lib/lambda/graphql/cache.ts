import { DeleteObjectsCommand, GetObjectCommand, ObjectIdentifier, PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { getEnvironment } from '../environment';
import { log } from '../logger';
import { Readable } from 'stream';

const s3Client = new S3Client({ region: 'eu-west-1' });

const AIS_LOCATION_CACHE = {
  MAX_AGE: 2, // 2 seconds
  STALE_WHILE_REVALIDATE: 3, // 3 seconds
  STALE_IF_ERROR: 58, // 58 seconds (do not return location data more than 1 minute old)
};
const AIS_VESSEL_CACHE = {
  MAX_AGE: 300, // 5 minutes
  STALE_WHILE_REVALIDATE: 60, // 1 minute
  STALE_IF_ERROR: 300, // 5 minutes (do not return vessel data more than 10 minutes old)
};

export const FEATURE_CACHE_DURATION = 7200; // 2 hours

function getCacheBucketName() {
  return `featurecache-${getEnvironment()}`;
}

export function getFeatureCacheDuration(key: string) {
  if (key === 'aislocations') {
    return AIS_LOCATION_CACHE.MAX_AGE;
  } else if (key === 'aisvessels') {
    return AIS_VESSEL_CACHE.MAX_AGE;
  } else {
    return FEATURE_CACHE_DURATION;
  }
}

export function getCacheControlHeaders(key: string): Record<string, string[]> {
  if (key === 'aislocations') {
    return {
      'Cache-Control': [
        'max-age=' +
          AIS_LOCATION_CACHE.MAX_AGE +
          ', ' +
          'stale-while-revalidate=' +
          AIS_LOCATION_CACHE.STALE_WHILE_REVALIDATE +
          ', ' +
          'stale-if-error=' +
          AIS_LOCATION_CACHE.STALE_IF_ERROR,
      ],
    };
  } else if (key === 'aisvessels') {
    return {
      'Cache-Control': [
        'max-age=' +
          AIS_VESSEL_CACHE.MAX_AGE +
          ', ' +
          'stale-while-revalidate=' +
          AIS_VESSEL_CACHE.STALE_WHILE_REVALIDATE +
          ', ' +
          'stale-if-error=' +
          AIS_VESSEL_CACHE.STALE_IF_ERROR,
      ],
    };
  } else {
    return {};
  }
}

export async function cacheResponse(key: string, response: object | string) {
  const cacheDurationSeconds = getFeatureCacheDuration(key);
  const expires = new Date();
  expires.setTime(expires.getTime() + cacheDurationSeconds * 1000);
  const command = new PutObjectCommand({
    Key: key,
    Bucket: getCacheBucketName(),
    Expires: expires,
    Body: typeof response === 'string' ? response : JSON.stringify(response),
  });
  await s3Client.send(command);
}

export type CacheResponse = {
  expired: boolean;
  data?: string;
  lastModified?: string;
};

export async function getFromCache(key: string): Promise<CacheResponse> {
  try {
    const data = await s3Client.send(
      new GetObjectCommand({
        Key: key,
        Bucket: getCacheBucketName(),
      })
    );
    if (data.Body) {
      log.debug(`returning ${key} from cache`);
      const streamToString = (stream: Readable): Promise<string> =>
        new Promise((resolve, reject) => {
          const chunks: Uint8Array[] = [];
          stream.on('data', (chunk: Uint8Array) => chunks.push(chunk));
          stream.on('error', reject);
          stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
        });
      const expires = data.ExpiresString ? Date.parse(data.ExpiresString) : undefined;
      return {
        expired: expires !== undefined && !isNaN(expires) && expires < Date.now(),
        data: await streamToString(data.Body as Readable),
        lastModified: data.LastModified?.toUTCString(),
      };
    }
  } catch (e) {
    // errors ignored also not found
  }
  return { expired: true };
}

export async function deleteCacheObjects(keys: string[]) {
  const objects: ObjectIdentifier[] = keys.map((key) => {
    return { Key: key };
  });
  const response = await s3Client.send(
    new DeleteObjectsCommand({
      Bucket: getCacheBucketName(),
      Delete: {
        Objects: objects,
      },
    })
  );
  log.debug(`Deleted cache objects ${keys}: ${response.Deleted?.length}/${keys.length}`);
  if (response.Errors?.length) {
    log.error(response.Errors);
  }
}
