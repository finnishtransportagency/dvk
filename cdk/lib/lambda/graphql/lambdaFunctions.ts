import * as path from 'path';

interface BackendLambda {
  entry: string;
  fieldName: string; // this should correlate to graphql schema field name
  typeName: string;
}

const lambdaFunctions: BackendLambda[] = [
  {
    entry: path.join(__dirname, 'query/fairways-handler.ts'),
    typeName: 'Query',
    fieldName: 'fairways',
  },
  {
    entry: path.join(__dirname, 'query/fairway-handler.ts'),
    typeName: 'Query',
    fieldName: 'fairway',
  },
  {
    entry: path.join(__dirname, 'query/fairwayNavigationLines-handler.ts'),
    typeName: 'Fairway',
    fieldName: 'navigationLines',
  },
  {
    entry: path.join(__dirname, 'query/navigationLines-handler.ts'),
    typeName: 'Query',
    fieldName: 'navigationLines',
  },
];

export default lambdaFunctions;
