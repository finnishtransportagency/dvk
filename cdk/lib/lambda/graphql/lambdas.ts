import * as path from 'path';
interface BackendLambda {
  entry: string;
  functionName: string;
  typeName: 'Query' | 'Mutation';
}

const lambdaFunctions: BackendLambda[] = [
  {
    entry: path.join(__dirname, 'query/fairways-handler.ts'),
    functionName: 'fairways',
    typeName: 'Query',
  },
];

export default lambdaFunctions;
