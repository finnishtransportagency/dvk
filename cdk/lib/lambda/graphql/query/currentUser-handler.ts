import { AppSyncResolverEvent } from 'aws-lambda/trigger/appsync-resolver';
import { CurrentUser } from '../../../../graphql/generated';
import { validateJwtToken } from '../../api/login';
import { log } from '../../logger';

export const handler = async (event: AppSyncResolverEvent<void>): Promise<CurrentUser> => {
  log.info({ headers: event.request.headers }, `currentUser(${event.identity})`);
  const token = event.request.headers['x-iam-accesstoken'];
  const data = event.request.headers['x-iam-data'];
  const jwtToken = await validateJwtToken(token, data || '');
  log.debug('JwtToken: %s', jwtToken);
  return {
    name: jwtToken ? jwtToken['custom:etunimi'] + ' ' + jwtToken['custom:sukunimi'] : '????',
  };
};
