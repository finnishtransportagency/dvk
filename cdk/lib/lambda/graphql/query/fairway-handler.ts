import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda/trigger/appsync-resolver';
import { Fairway, QueryFairwayArgs } from '../../../../graphql/generated';
import { log } from '../../logger';
import FairwayDBModel from '../../db/fairwayModel';
import { FairwayService } from '../../api/fairwayService';

const fairwayService = new FairwayService();

export const handler: AppSyncResolverHandler<QueryFairwayArgs, Fairway> = async (event: AppSyncResolverEvent<QueryFairwayArgs>): Promise<Fairway> => {
  log.info(`fairway(${event.arguments.id})`);
  const dbModel = await FairwayDBModel.get(event.arguments.id);
  const apiModel = fairwayService.getFairway(event.arguments.id);
  const fairway: Fairway = {
    id: event.arguments.id,
    name: {
      fi: apiModel.nimiFI,
      sv: apiModel.nimiSV || '',
      en: dbModel?.name || '',
    },
    draft1: apiModel.kulkuSyvyys1,
    draft2: apiModel.kulkuSyvyys2,
    draft3: apiModel.kulkuSyvyys3,
    length: apiModel.pituus,
  };
  return fairway;
};
