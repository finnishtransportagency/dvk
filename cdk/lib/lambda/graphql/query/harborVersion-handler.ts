import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda/trigger/appsync-resolver';
import { Harbor, QueryHarborArgs } from '../../../../graphql/generated';
import { log } from '../../logger';
import HarborDBModel from '../../db/harborDBModel';
import { mapHarborDBModelToGraphqlType } from '../../db/modelMapper';
import { getOptionalCurrentUser } from '../../api/login';

export const handler: AppSyncResolverHandler<QueryHarborArgs, Harbor | undefined> = async (
  event: AppSyncResolverEvent<QueryHarborArgs>
): Promise<Harbor | undefined> => {
  log.info(`harbor(${event.arguments.id})`);
  const user = await getOptionalCurrentUser(event);
  const harbor = await HarborDBModel.get(event.arguments.id, event.arguments.version);
  return harbor ? mapHarborDBModelToGraphqlType(harbor, user) : undefined;
};