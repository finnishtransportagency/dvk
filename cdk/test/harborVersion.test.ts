import { mockClient } from 'aws-sdk-client-mock';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';
import { handler } from '../lib/lambda/graphql/query/harborVersion-handler';
import HarborDBModel from '../lib/lambda/db/harborDBModel';
import { Status } from '../graphql/generated';
import { mockContext, mockQueryByIdAndVersionEvent, mockQueryByIdEvent } from './mocks';

const harbor1: HarborDBModel = {
  id: 'test',
  version: 'v1',
  name: {
    fi: 'Harbor1fi',
    sv: 'Harbor1sv',
    en: 'Harbor1en',
  },
  creator: 'test',
  creationTimestamp: Date.now(),
  modifier: 'test2',
  modificationTimestamp: Date.now(),
  n2000HeightSystem: true,
  geometry: { coordinates: [18, 62] },
  status: Status.Public,
  cargo: {
    fi: 'cargofi',
    sv: 'cargosv',
    en: 'cargoen',
  },
  company: {
    fi: 'companyfi',
    sv: 'companysv',
    en: 'companyen',
  },
  email: 'email@fi',
  extraInfo: {
    fi: 'extrafi',
    sv: 'extrasv',
    en: 'extraen',
  },
  fax: '32321',
  internet: 'www.fi',
  phoneNumber: ['12345', '22222'],
  harborBasin: {
    fi: 'basinfi',
    sv: 'basinsv',
    en: 'basinen',
  },
  quays: [
    {
      extraInfo: {
        fi: 'cargofi',
        sv: 'cargosv',
        en: 'cargoen',
      },
      name: {
        fi: 'cargofi',
        sv: 'cargosv',
        en: 'cargoen',
      },
      geometry: { coordinates: [22.1, 63.2] },
      length: 123,
      sections: [
        {
          depth: 10,
          geometry: { coordinates: [17, 62.2] },
          name: 'stest',
        },
      ],
    },
  ],
};

const harbor2: HarborDBModel = {
  id: 'test',
  version: 'v2',
  name: {
    fi: 'Harbor2fi',
    sv: 'Harbor2sv',
    en: 'Harbor2en',
  },
  creator: 'test',
  creationTimestamp: Date.now(),
  modifier: 'test2',
  modificationTimestamp: Date.now(),
  status: Status.Draft,
};

const ddbMock = mockClient(DynamoDBDocumentClient);

jest.mock('../lib/lambda/environment', () => ({
  getEnvironment: () => 'mock',
  isPermanentEnvironment: () => false,
  getHarborTableName: () => 'Harbor-mock',
}));

beforeEach(() => {
  ddbMock.reset();
});

it('should get v1 by id from the DynamoDB', async () => {
  ddbMock
    .on(GetCommand, {
      Key: { id: 'test' },
    })
    .resolves({
      Item: harbor1,
    });
  const response = await handler(mockQueryByIdEvent, mockContext, () => {});
  expect(response).toMatchSnapshot({
    modificationTimestamp: expect.any(Number),
    creationTimestamp: expect.any(Number),
  });
});

it('should get harbor by id and version from the DynamoDB', async () => {
  ddbMock
    .on(GetCommand, {
      Key: { id: 'test', version: 'v2' },
    })
    .resolves({
      Item: harbor2,
    });
  const response = await handler(mockQueryByIdAndVersionEvent, mockContext, () => {});
  expect(response).toMatchSnapshot({
    modificationTimestamp: expect.any(Number),
    creationTimestamp: expect.any(Number),
  });
});

it('should get undefined when version not present', async () => {
  ddbMock
    .on(GetCommand, {
      Key: { id: 'test', version: 'v5' },
    })
    .resolves({
      Item: harbor2,
    });
  const response = await handler(mockQueryByIdAndVersionEvent, mockContext, () => {});
  expect(response).toBe(undefined);
});
