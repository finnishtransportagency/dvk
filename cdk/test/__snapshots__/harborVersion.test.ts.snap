// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get harbor by id and version from the DynamoDB 1`] = `
{
  "cargo": undefined,
  "company": undefined,
  "creationTimestamp": Any<Number>,
  "creator": "test",
  "email": undefined,
  "extraInfo": undefined,
  "fax": undefined,
  "geometry": undefined,
  "harborBasin": undefined,
  "id": "test",
  "internet": undefined,
  "modificationTimestamp": Any<Number>,
  "modifier": "test2",
  "n2000HeightSystem": undefined,
  "name": {
    "en": "Harbor2en",
    "fi": "Harbor2fi",
    "sv": "Harbor2sv",
  },
  "phoneNumber": undefined,
  "quays": undefined,
  "status": "DRAFT",
  "version": "v2",
}
`;

exports[`should get v1 by id from the DynamoDB 1`] = `
{
  "cargo": {
    "en": "cargoen",
    "fi": "cargofi",
    "sv": "cargosv",
  },
  "company": {
    "en": "companyen",
    "fi": "companyfi",
    "sv": "companysv",
  },
  "creationTimestamp": Any<Number>,
  "creator": "test",
  "email": "email@fi",
  "extraInfo": {
    "en": "extraen",
    "fi": "extrafi",
    "sv": "extrasv",
  },
  "fax": "32321",
  "geometry": {
    "coordinates": [
      18,
      62,
    ],
  },
  "harborBasin": {
    "en": "basinen",
    "fi": "basinfi",
    "sv": "basinsv",
  },
  "id": "test",
  "internet": "www.fi",
  "modificationTimestamp": Any<Number>,
  "modifier": "test2",
  "n2000HeightSystem": true,
  "name": {
    "en": "Harbor1en",
    "fi": "Harbor1fi",
    "sv": "Harbor1sv",
  },
  "phoneNumber": [
    "12345",
    "22222",
  ],
  "quays": [
    {
      "extraInfo": {
        "en": "cargoen",
        "fi": "cargofi",
        "sv": "cargosv",
      },
      "geometry": {
        "coordinates": [
          22.1,
          63.2,
        ],
      },
      "length": 123,
      "name": {
        "en": "cargoen",
        "fi": "cargofi",
        "sv": "cargosv",
      },
      "sections": [
        {
          "depth": 10,
          "geometry": {
            "coordinates": [
              17,
              62.2,
            ],
          },
          "name": "stest",
        },
      ],
    },
  ],
  "status": "PUBLIC",
  "version": "v1",
}
`;
